### PR TITLE
feature/triptimes-graphql-api

### PR DESCRIPTION
**Pull request by kenkus-futurice**
GraphQL endpoint for getting cancelled TripTimes

 - Filters cancelled TripTimes by feed ID
 - Filters cancelled TripTimes by specific date
 - Filters cancelled TripTimes by scheduled last stop arrival on or after specific date
 - Filters cancelled TripTimes by scheduled last stop arrival on or after specific time

Screenshot 2019-03-26 at 18 31 49

### CODE REVIEW

**Comment by kenkus-futurice** [comment]
Not sure how to override `hashCode()` and `equals()` in `TripTimeShort`. Or if that is a good idea in the first place.

File: src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java [hunk 1]
```diff
@@ -3054,6 +3055,55 @@ private Object getObject(String idString) {
...
+                        .type(new GraphQLList(stoptimeType))
+                        .dataFetcher(environment -> {
+                            final String feedId = environment.getArgument("feedId");
+                            ServiceDate maybeServiceDate = null;
+                            try {
+                                maybeServiceDate = ServiceDate.parseString(environment.getArgument("serviceDate"));
+                            } catch (Exception e) {
+
+                            }
+                            final ServiceDate serviceDate = maybeServiceDate;
```
Comment (pailakka) [inline src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java]: 3080 - 3104 should probably be a seperate method in GraphIndex

File: src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java [hunk 2]
```diff
@@ -3054,6 +3055,55 @@ private Object getObject(String idString) {
...
+                                .type(Scalars.GraphQLString)
+                                .build())
+                        .description("Get cancelled TripTimes.")
+                        .type(new GraphQLList(stoptimeType))
+                        .dataFetcher(environment -> {
+                            final String feedId = environment.getArgument("feedId");
+                            ServiceDate maybeServiceDate = null;
+                            try {
+                                maybeServiceDate = ServiceDate.parseString(environment.getArgument("serviceDate"));
+                            } catch (Exception e) {
```
Comment (pailakka) [inline src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java]: We should catch only `ParseException`

File: src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java [hunk 3]
```diff
@@ -3054,6 +3055,55 @@ private Object getObject(String idString) {
...
+                                    .filter(timetable -> (serviceDate != null) ? serviceDate.equals(timetable.serviceDate) : timetable.serviceDate != null)
+                                    .flatMap(timetable -> timetable.tripTimes
+                                            .stream()
+                                            .filter(tripTimes -> tripTimes.getRealTimeState() == RealTimeState.CANCELED)
+                                            .map(tripTimes -> {
+                                                final int stopIndex = 0;
+                                                final String agencyId = tripTimes.trip.getId().getAgencyId();
+                                                final Stop stop = timetable.pattern.getStop(stopIndex);
+                                                final CalendarService calendarService = index.graph.getCalendarService();
+                                                final ServiceDay serviceDay = new ServiceDay(index.graph, timetable.serviceDate, calendarService, agencyId);
```
Comment (pailakka) [inline src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java]: Creating unnecessary `ServiceDay` objects, which is slow

File: src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java (hunk 1 above)
Comment (kenkus-futurice) [reply src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java]: Good point.

File: src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java (hunk 2 above)
Comment (kenkus-futurice) [reply src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java]: Added also `NullPointerException`

File: src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java (hunk 3 above)
Comment (kenkus-futurice) [reply src/main/java/org/opentripplanner/index/IndexGraphQLSchema.java]: Nice improvement

**Comment by kenkus-futurice** [comment]
Added new filtering features. Please review. Thanks!

**Pull request merged**